lib/src/rules/avoid_type_to_string.dart

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+
+import '../analyzer.dart';
+
+const _desc =
+    r'Avoid <Type>.toString() in production code since results may be minified.';
+
+const _details = r'''
+
+**DO** avoid calls to <Type>.toString() in production code, since it does not
+contractually return the user-defined name of the Type (or underlying class).
+Development-mode compilers where code size is not a concern use the full name,
+but release-mode compilers often choose to minify these symbols.
+
+**BAD:**
+```
+void bar(Object other) {
+  if (other.runtimeType.toString() == 'Bar') {
+    doThing();
+  }
+}
+
+Object baz(Thing myThing) {
+  return getThingFromDatabase(key: myThing.runtimeType.toString());
+}
+```
+
+**GOOD:**
+```
+void bar(Object other) {
+  if (other is Bar) {
+    doThing();
+  }
+}
+
+class Thing {
+  String get thingTypeKey => ...
+}
+
+Object baz(Thing myThing) {
+  return getThingFromDatabase(key: myThing.thingTypeKey);
+}
+```
+
+''';
+
+class AvoidTypeToString extends LintRule implements NodeLintRule {
+  AvoidTypeToString()
+      : super(
+            name: 'avoid_type_to_string',
+            description: _desc,
+            details: _details,
+            group: Group.errors);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    assert(context != null);
+    final visitor =
+        _Visitor(this, context.typeSystem, context.typeProvider.typeType);
+    // Gathering meta information at these nodes.
+    // Nodes visited in DFS, so this will be called before
+    // each visitMethodInvocation.
+    registry.addClassDeclaration(this, visitor);
+    registry.addMixinDeclaration(this, visitor);
+    registry.addExtensionDeclaration(this, visitor);
+
+    // These nodes delegate to general visitArgumentList,
+    // since SimpleAstVisitor only calls visits for concrete node subtypes.
```

Comment:
Sorry, I should have been more explicit. What I was thinking is that this code could register to visit argument lists:
```
registry.addArgumentList(this, visitor);
```
and you could process all argument lists by overriding `visitArgumentList` (which you are now already doing). In other words, that one line could replace lines 77-80. This has the advantage that if a new language construct is added and the AST gets a new node containing an argument list, this lint will continue to work without needing to be updated.

Reply:
> This has the advantage that if a new language construct is added and the AST gets a new node containing an argument list, this lint will continue to work without needing to be updated.

I had the same concern, but I wasn't able to find an `addArgumentList` method on `registry`, or [anywhere else in dart-lang/sdk](https://github.com/dart-lang/sdk/search?q=addArgumentList&unscoped_q=addArgumentList). The closest option seems to be `addTypeArgumentList`, but that seems to be for things like function declarations?

Reply:
You're right, that node appears to have been missed when writing the registry API. I'll get that fixed, but it will be a while before it's published and available for the linter package to use. In the meantime, this code is the best we can do, but a TODO comment suggesting this change would be good.

Reply:
👍